skimage/segmentation/_watershed_cy.pyx

```diff
@@ -117,13 +121,17 @@ def watershed_raveled(cnp.float64_t[::1] image,
     cdef Py_ssize_t index = 0
     cdef Py_ssize_t neighbor_index = 0
     cdef DTYPE_BOOL_t compact = (compactness > 0)
+    cdef cnp.float64_t neg_inf = -np.inf
 
     cdef Heap *hp = <Heap *> heap_from_numpy2()
 
     with nogil:
         for i in range(marker_locations.shape[0]):
             index = marker_locations[i]
-            elem.value = image[index]
+            if from_minima:
+                elem.value = image[index]
+            else:
+                elem.value = neg_inf
```

Comment:
I am still confused about why we need to have this test for `from_minima`. Why not just set
```suggestion
            elem.value = neg_inf
```
always. If markers are identical with local minima, the marker should already be smaller than its neighbors.

I tried this assumption locally but https://github.com/scikit-image/scikit-image/blob/89fa48aabdd96d7b6ae47dd111393564b7488384/skimage/segmentation/tests/test_watershed.py#L515

fails with `assert 1300 == 1301`. I am not sure yet why. :thinking:

Reply:
Hmm, I think this may be because with `elem.value = -np.inf` the push and pop order of the markers is identical, while with `elem.value = image[index]` id depends on the markers value. 

My gut feeling is telling me that we don't want this, because now the output of `watershed(..., markers, ...)` also depends entirely on the given order of the `markers`. Additionally, if the user provides the minima themselves, they may get a different result compared to passing `markers=None` even though the docstring tells them that the result should be the same!

Going from the source cited in https://github.com/scikit-image/scikit-image/pull/7071#discussion_r1271382091, I tend to prefer the following approach: set `elem.value = neg_inf` for all markers and sort the initial marker order by their image value. If the performance penalty due to sorting should be avoided, I am fine with just using the order of given markers, but that should be clearly documented in the docstring. What do you think?

Reply:
Oh, and I tested this, the [failing test](https://github.com/scikit-image/scikit-image/blob/89fa48aabdd96d7b6ae47dd111393564b7488384/skimage/segmentation/tests/test_watershed.py#L515) also fails if the markers are sorted by their value. I assume that the difference is due to similar behavior as in https://github.com/scikit-image/scikit-image/issues/6632...